Unable to (re)start WebsocketManager on screen on due to background restrictions

<!-- READ THIS FIRST:
- Make sure you run the latest version of the Android app
- Make sure you run the latest version of Home Assistant
- Make sure to check the Companion docs for troubleshooting and configuration: https://companion.home-assistant.io/
- Make sure the bug you found is not already reported, we love to put work in bugfixes instead of closing duplicate bug reports
  DO NOT DELETE ANY TEXT from this template! All requested information is important.
-->

**Home Assistant Android version:** beta-1944-6994eacf

**Android version:** 12

**Phone model:** Google Pixel 4a

**Home Assistant version:** Core 2021.12.10

**Last working Home Assistant release (if known):** n/a, new feature

**Description of problem:** I upgraded to the beta version directly from the latest stable release (2022.1.1). No settings changed for websocket notifications (so: connect while screen is on). I noticed that the notification that the app was connected didn't come up after picking up my phone again. Logcat shows an error (see below) which is caused by the [Android 12 background start restrictions](https://developer.android.com/guide/components/foreground-services#background-start-restrictions).

**Traceback (if applicable, to get the logs you may refer to: https://companion.home-assistant.io/docs/troubleshooting/faqs/#android-crash-logs):**

Slightly edited to start when the screen turns on, and to include a message from the ActivityManager

```
2022-01-19 21:04:29.999 9988-9988/? D/SensorReceiver: Sensor is_interactive corresponding to received event android.intent.action.SCREEN_ON is disabled, skipping sensors update
2022-01-19 21:04:30.002 9988-9988/? D/ButtonWidget: Broadcast received: 
    Broadcast action: android.intent.action.SCREEN_ON
    AppWidgetId: -1
2022-01-19 21:04:30.002 9988-9988/? D/ButtonWidget: Updating all widgets
2022-01-19 21:04:30.048 9988-9988/? D/UrlRepository: localUrl is: true and usesInternalSsid is: true
2022-01-19 21:04:30.058 9988-9988/? D/MediaPlayCtrlsWidget: Broadcast received: 
    Broadcast action: android.intent.action.SCREEN_ON
    AppWidgetId: -1
2022-01-19 21:04:30.063 9988-9988/? D/UrlRepository: localUrl is: true and usesInternalSsid is: true
2022-01-19 21:04:30.071 9988-9988/? D/UrlRepository: localUrl is: true and usesInternalSsid is: true
2022-01-19 21:04:30.093 9988-10057/? D/WebSockManager: Starting to listen to Websocket
2022-01-19 21:04:30.098 9988-10073/? I/WM-Processor: Moving WorkSpec (2698abf6-4e72-456d-b0bd-09ec4047eb87) to the foreground
2022-01-19 21:04:30.114 9988-10069/? E/WM-WorkerWrapper: Work [ id=2698abf6-4e72-456d-b0bd-09ec4047eb87, tags={ io.homeassistant.companion.android.websocket.WebsocketManager } ] failed because it threw an exception/error
    java.util.concurrent.ExecutionException: android.app.ForegroundServiceStartNotAllowedException: startForegroundService() not allowed due to mAllowStartForeground false: service io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService
        at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:516)
        at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
        at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:311)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
     Caused by: android.app.ForegroundServiceStartNotAllowedException: startForegroundService() not allowed due to mAllowStartForeground false: service io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService
        at android.app.ForegroundServiceStartNotAllowedException$1.createFromParcel(ForegroundServiceStartNotAllowedException.java:54)
        at android.app.ForegroundServiceStartNotAllowedException$1.createFromParcel(ForegroundServiceStartNotAllowedException.java:50)
        at android.os.Parcel.readParcelable(Parcel.java:3333)
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2420)
        at android.os.Parcel.createException(Parcel.java:2409)
        at android.os.Parcel.readException(Parcel.java:2392)
        at android.os.Parcel.readException(Parcel.java:2334)
        at android.app.IActivityManager$Stub$Proxy.startService(IActivityManager.java:5971)
        at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1847)
        at android.app.ContextImpl.startForegroundService(ContextImpl.java:1823)
        at android.content.ContextWrapper.startForegroundService(ContextWrapper.java:779)
        at androidx.core.content.ContextCompat$Api26Impl.startForegroundService(ContextCompat.java:931)
        at androidx.core.content.ContextCompat.startForegroundService(ContextCompat.java:703)
        at androidx.work.impl.Processor.startForeground(Processor.java:161)
        at androidx.work.impl.utils.WorkForegroundUpdater$1.run(WorkForegroundUpdater.java:100)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:920) 
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.am.ActiveServices.startServiceLocked(ActiveServices.java:691)
        at com.android.server.am.ActiveServices.startServiceLocked(ActiveServices.java:616)
        at com.android.server.am.ActivityManagerService.startService(ActivityManagerService.java:11850)
        at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:2519)
        at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2498)
2022-01-19 21:04:30.115 9988-10069/? I/WM-WorkerWrapper: Worker result FAILURE for Work [ id=2698abf6-4e72-456d-b0bd-09ec4047eb87, tags={ io.homeassistant.companion.android.websocket.WebsocketManager } ]

2022-01-19 21:04:30.118 1740-2686/? W/ActivityManager: Background start not allowed: service Intent { act=ACTION_STOP_FOREGROUND cmp=io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService } to io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService from pid=9988 uid=10365 pkg=io.homeassistant.companion.android startFg?=false

2022-01-19 21:04:30.118 9988-10069/? E/WM-Processor: Unable to stop foreground service
    android.app.BackgroundServiceStartNotAllowedException: Not allowed to start service Intent { act=ACTION_STOP_FOREGROUND cmp=io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService }: app is in background uid UidRecord{f43bc62 u0a365 TRNB idle procs:0 seq(0,0,0)}
        at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1861)
        at android.app.ContextImpl.startService(ContextImpl.java:1817)
        at android.content.ContextWrapper.startService(ContextWrapper.java:774)
        at androidx.work.impl.Processor.stopForegroundService(Processor.java:318)
        at androidx.work.impl.Processor.stopForeground(Processor.java:224)
        at androidx.work.impl.WorkerWrapper.resolve(WorkerWrapper.java:460)
        at androidx.work.impl.WorkerWrapper.resetPeriodicAndResolve(WorkerWrapper.java:571)
        at androidx.work.impl.WorkerWrapper.handleResult(WorkerWrapper.java:490)
        at androidx.work.impl.WorkerWrapper.onWorkFinished(WorkerWrapper.java:354)
        at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:331)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
```

**Screenshot of problem:**

**Additional information:**

--- a/app/src/main/java/io/homeassistant/companion/android/settings/websocket/WebsocketSettingFragment.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/settings/websocket/WebsocketSettingFragment.kt
@@ -2,13 +2,21 @@ package io.homeassistant.companion.android.settings.websocket
 
 import android.content.Intent
 import android.net.Uri
+import android.os.Build
 import android.os.Bundle
+import android.os.PowerManager
+import android.provider.Settings
 import android.view.LayoutInflater
 import android.view.Menu
 import android.view.View
 import android.view.ViewGroup
+import androidx.activity.result.contract.ActivityResultContracts
 import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
 import androidx.compose.ui.platform.ComposeView
+import androidx.core.content.getSystemService
 import androidx.fragment.app.Fragment
 import androidx.fragment.app.viewModels
 import com.google.android.material.composethemeadapter.MdcTheme
@@ -22,6 +30,12 @@ import io.homeassistant.companion.android.common.R as commonR
 class WebsocketSettingFragment : Fragment() {
 
     val viewModel: SettingViewModel by viewModels()
+
+    private var isIgnoringBatteryOptimizations by mutableStateOf(false)
+
+    private val requestBackgroundAccessResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+        setIgnoringBatteryOptimizations()
+    }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
@@ -49,7 +63,18 @@ class WebsocketSettingFragment : Fragment() {
                         .collectAsState(initial = viewModel.getSetting(0))
                     WebsocketSettingView(
                         websocketSetting = settings.value.websocketSetting,
-                        onSettingChanged = { viewModel.updateWebsocketSetting(0, it) }
+                        unrestrictedBackgroundAccess = isIgnoringBatteryOptimizations,
+                        onSettingChanged = { viewModel.updateWebsocketSetting(0, it) },
+                        onBackgroundAccessTapped = {
+                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                                requestBackgroundAccessResult.launch(
+                                    Intent(
+                                        Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS,
+                                        Uri.parse("package:${activity?.packageName}")
+                                    )
+                                )
+                            }
+                        }
                     )
                 }
             }
@@ -59,5 +84,13 @@ class WebsocketSettingFragment : Fragment() {
     override fun onResume() {
         super.onResume()
         activity?.title = getString(commonR.string.websocket_setting_name)
+        setIgnoringBatteryOptimizations()
+    }
+
+    private fun setIgnoringBatteryOptimizations() {
+        isIgnoringBatteryOptimizations = Build.VERSION.SDK_INT <= Build.VERSION_CODES.M ||
+            context?.getSystemService<PowerManager>()
+                ?.isIgnoringBatteryOptimizations(requireActivity().packageName)
+                ?: false
     }
 }

--- a/app/src/main/java/io/homeassistant/companion/android/settings/websocket/views/WebsocketSettingView.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/settings/websocket/views/WebsocketSettingView.kt
@@ -5,6 +5,8 @@ import android.content.res.Configuration
 import android.os.Build
 import androidx.compose.foundation.layout.Box
 import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Spacer
+import androidx.compose.foundation.layout.height
 import androidx.compose.foundation.layout.padding
 import androidx.compose.foundation.rememberScrollState
 import androidx.compose.foundation.verticalScroll
@@ -20,13 +22,16 @@ import io.homeassistant.companion.android.BuildConfig
 import io.homeassistant.companion.android.common.R
 import io.homeassistant.companion.android.common.util.websocketChannel
 import io.homeassistant.companion.android.database.settings.WebsocketSetting
+import io.homeassistant.companion.android.util.compose.HaAlertWarning
 import io.homeassistant.companion.android.util.compose.InfoNotification
 import io.homeassistant.companion.android.util.compose.RadioButtonRow
 
 @Composable
 fun WebsocketSettingView(
     websocketSetting: WebsocketSetting,
-    onSettingChanged: (WebsocketSetting) -> Unit
+    unrestrictedBackgroundAccess: Boolean,
+    onSettingChanged: (WebsocketSetting) -> Unit,
+    onBackgroundAccessTapped: () -> Unit
 ) {
     val scrollState = rememberScrollState()
     val context = LocalContext.current
@@ -36,6 +41,14 @@ fun WebsocketSettingView(
                 text = stringResource(R.string.websocket_setting_description),
                 modifier = Modifier.padding(bottom = 16.dp)
             )
+            if (!unrestrictedBackgroundAccess && websocketSetting != WebsocketSetting.NEVER) {
+                HaAlertWarning(
+                    message = stringResource(R.string.websocket_notification_backgroundaccess),
+                    action = stringResource(R.string.allow),
+                    onActionClicked = onBackgroundAccessTapped
+                )
+                Spacer(modifier = Modifier.height(16.dp))
+            }
             Divider()
             RadioButtonRow(
                 text = stringResource(if (BuildConfig.FLAVOR == "full") R.string.websocket_setting_never else R.string.websocket_setting_never_minimal),

--- a/app/src/main/java/io/homeassistant/companion/android/util/compose/HaAlert.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/util/compose/HaAlert.kt
@@ -1,0 +1,46 @@
+package io.homeassistant.companion.android.util.compose
+
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material.ButtonDefaults
+import androidx.compose.material.MaterialTheme
+import androidx.compose.material.Text
+import androidx.compose.material.TextButton
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.colorResource
+import androidx.compose.ui.unit.dp
+import io.homeassistant.companion.android.common.R as commonR
+
+@Composable
+fun HaAlertWarning(
+    message: String,
+    action: String?,
+    onActionClicked: () -> Unit
+) {
+    Row(
+        modifier = Modifier
+            .fillMaxWidth()
+            .background(colorResource(commonR.color.colorAlertWarning), MaterialTheme.shapes.medium)
+            .padding(all = 8.dp),
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        Text(
+            text = message,
+            modifier = Modifier
+                .weight(1f)
+                .padding(end = if (action != null) 8.dp else 0.dp)
+        )
+        if (action != null) {
+            TextButton(
+                colors = ButtonDefaults.textButtonColors(contentColor = colorResource(commonR.color.colorOnAlertWarning)),
+                onClick = onActionClicked
+            ) {
+                Text(action)
+            }
+        }
+    }
+}

--- a/app/src/main/java/io/homeassistant/companion/android/websocket/WebsocketManager.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/websocket/WebsocketManager.kt
@@ -27,6 +27,7 @@ import io.homeassistant.companion.android.common.R
 import io.homeassistant.companion.android.common.data.url.UrlRepository
 import io.homeassistant.companion.android.common.data.websocket.WebSocketRepository
 import io.homeassistant.companion.android.common.util.websocketChannel
+import io.homeassistant.companion.android.common.util.websocketIssuesChannel
 import io.homeassistant.companion.android.database.settings.SettingsDao
 import io.homeassistant.companion.android.database.settings.WebsocketSetting
 import io.homeassistant.companion.android.notifications.MessagingManager
@@ -37,6 +38,7 @@ import kotlinx.coroutines.delay
 import kotlinx.coroutines.isActive
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
+import java.lang.IllegalStateException
 import java.util.concurrent.TimeUnit
 
 class WebsocketManager(
@@ -48,6 +50,7 @@ class WebsocketManager(
         private const val TAG = "WebSockManager"
         private const val SOURCE = "Websocket"
         private const val NOTIFICATION_ID = 65423
+        private const val NOTIFICATION_RESTRICTED_ID = 65424
         private val DEFAULT_WEBSOCKET_SETTING = if (BuildConfig.FLAVOR == "full") WebsocketSetting.NEVER else WebsocketSetting.ALWAYS
 
         fun start(context: Context) {
@@ -99,7 +102,7 @@ class WebsocketManager(
         }
 
         Log.d(TAG, "Starting to listen to Websocket")
-        createNotification()
+        if (!createNotification()) return@withContext Result.success()
 
         // Start listening for notifications
         val job = launch { collectNotifications() }
@@ -163,7 +166,12 @@ class WebsocketManager(
         }
     }
 
-    private suspend fun createNotification() {
+    /**
+     * Create a notification to start the service as a foreground service.
+     *
+     * @return `true` if the foreground service was started
+     */
+    private suspend fun createNotification(): Boolean {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
             var notificationChannel =
                 notificationManager.getNotificationChannel(websocketChannel)
@@ -210,6 +218,31 @@ class WebsocketManager(
                 settingPendingIntent
             )
             .build()
-        setForeground(ForegroundInfo(NOTIFICATION_ID, notification))
+        return try {
+            setForeground(ForegroundInfo(NOTIFICATION_ID, notification))
+            true
+        } catch (e: IllegalStateException) {
+            Log.e(TAG, "Unable to setForeground due to restrictions", e)
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                if (notificationManager.getNotificationChannel(websocketIssuesChannel) == null) {
+                    val restrictedNotificationChannel = NotificationChannel(
+                        websocketIssuesChannel,
+                        applicationContext.getString(R.string.websocket_notification_issues),
+                        NotificationManager.IMPORTANCE_DEFAULT
+                    )
+                    notificationManager.createNotificationChannel(restrictedNotificationChannel)
+                }
+            }
+            val restrictedNotification = NotificationCompat.Builder(applicationContext, websocketIssuesChannel)
+                .setSmallIcon(R.drawable.ic_stat_ic_notification)
+                .setContentTitle(applicationContext.getString(R.string.websocket_restricted_title))
+                .setContentText(applicationContext.getString(R.string.websocket_restricted_fix))
+                .setContentIntent(settingPendingIntent)
+                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
+                .setAutoCancel(true)
+                .build()
+            notificationManager.notify(NOTIFICATION_RESTRICTED_ID, restrictedNotification)
+            false
+        }
     }
 }

--- a/app/src/main/java/io/homeassistant/companion/android/websocket/WebsocketManager.kt
+++ b/app/src/main/java/io/homeassistant/companion/android/websocket/WebsocketManager.kt
@@ -101,10 +101,12 @@ class WebsocketManager(
             return@withContext Result.success()
         }
 
-        Log.d(TAG, "Starting to listen to Websocket")
-        if (!createNotification()) return@withContext Result.success()
+        if (!createNotification()) {
+            return@withContext Result.success()
+        }
 
         // Start listening for notifications
+        Log.d(TAG, "Starting to listen to Websocket")
         val job = launch { collectNotifications() }
 
         // play ping pong to ensure we have a connection.

--- a/common/src/main/java/io/homeassistant/companion/android/common/util/AppNotifChannels.kt
+++ b/common/src/main/java/io/homeassistant/companion/android/common/util/AppNotifChannels.kt
@@ -3,6 +3,7 @@ package io.homeassistant.companion.android.common.util
 const val sensorWorkerChannel = "Sensor Worker"
 const val sensorCoreSyncChannel = "Sensor Sync"
 const val websocketChannel = "Websocket"
+const val websocketIssuesChannel = "Websocket Issues"
 const val highAccuracyChannel = "High accuracy location"
 const val databaseChannel = "App Database"
 const val locationDisabledChannel = "Location disabled"
@@ -13,6 +14,7 @@ val appCreatedChannels = listOf(
     sensorWorkerChannel,
     sensorCoreSyncChannel,
     websocketChannel,
+    websocketIssuesChannel,
     highAccuracyChannel,
     databaseChannel,
     locationDisabledChannel,

--- a/common/src/main/res/values/colors.xml
+++ b/common/src/main/res/values/colors.xml
@@ -25,4 +25,6 @@
     <color name="colorActionBarPopupBackground">@android:color/white</color>
     <color name="colorLaunchScreenBackground">#fafafa</color>
     <color name="colorCodeBackground">#f5f5f5</color>
+    <color name="colorAlertWarning">#1fffa600</color>
+    <color name="colorOnAlertWarning">#ffa600</color>
 </resources>

--- a/common/src/main/res/values/strings.xml
+++ b/common/src/main/res/values/strings.xml
@@ -9,6 +9,7 @@
     <string name="add_ssid_name_suggestion">Add %1$s</string>
     <string name="add_widget">Add widget</string>
     <string name="all_entities">All entities</string>
+    <string name="allow">Allow</string>
     <string name="app_name">Home Assistant</string>
     <string name="app_version_info">App Version Info</string>
     <string name="application_version">Application Version</string>
@@ -748,6 +749,8 @@
     <string name="wear_set_favorites">Select your favorite entities to appear at the top of the Wear home screen. You can also drag and drop to change the order in which they appear.</string>
     <string name="wear_settings">Wear Device Settings</string>
     <string name="websocket_listening">Connected to Home Assistant</string>
+    <string name="websocket_restricted_title">Unable to open connection to Home Assistant</string>
+    <string name="websocket_restricted_fix">Update settings to fix or disable</string>
     <string name="webview_error_AUTH_SCHEME">Unsupported authentication scheme (not basic or digest), please check network settings.</string>
     <string name="webview_error_AUTHENTICATION">User authentication failed on server, please check server settings.</string>
     <string name="webview_error_description">Encountered error :</string>
@@ -829,7 +832,9 @@
     <string name="websocket_setting_always">Always</string>
     <string name="websocket_setting_always_minimal">Always\n\nRequired to always receive notifications</string>
     <string name="websocket_persistent_notification">In order to maintain the persistent connection the app will need to create a persistent notification. You may use the button below to manage the appearance of this notification. It is recommended to minimize the notification to hide the icon.</string>
+    <string name="websocket_notification_backgroundaccess">To reliably connect your server, allow Home Assistant to keep running in the background.</string>
     <string name="websocket_notification_channel">Manage Persistent Connection Notification</string>
+    <string name="websocket_notification_issues">Persistent Connection Issues</string>
     <string name="notification_channels">Notification Channels</string>
     <string name="notification_channels_summary">Manage all notification channels configured on the device. Channels control the behavior of its notifications including visibility and sound.</string>
     <string name="info">Information</string>

